weave/trace/weave_client_send_file_cache.py

```diff
@@ -0,0 +1,171 @@
+"""
+In memory cache of files sent to the server to avoid sending the same file
+multiple times, using an LRU (Least Recently Used) eviction policy.
+"""
+
+import hashlib
+import threading
+from collections import OrderedDict
+from concurrent.futures import Future
+from typing import Generic, Optional, TypeVar
+
+from weave.trace_server.trace_server_interface import FileCreateReq, FileCreateRes
+
+# Define generic type variables
+K = TypeVar("K")  # Key type
+V = TypeVar("V")  # Value type
+
+
+class ThreadSafeLRUCache(Generic[K, V]):
+    """Thread-safe LRU cache implementation using OrderedDict."""
+
+    def __init__(self, max_size: int = 1000) -> None:
+        """Initialize an LRU cache with a maximum size.
+
+        Args:
+            max_size: Maximum number of items to store in the cache.
+                     If set to 0 or negative, cache size is unlimited.
+        """
+        self._max_size = max(0, max_size)  # Ensure max_size is non-negative
+        self._cache: OrderedDict[K, V] = OrderedDict()
+        self._lock = threading.RLock()
+
+    def put(self, key: K, value: V) -> None:
+        """Thread-safe insertion/update with LRU tracking.
+
+        If the key already exists, it will be moved to the end (most recently used position).
+        If max_size is reached, the least recently used item will be evicted.
+        """
+        with self._lock:
+            # If key exists, remove it so we can re-insert at the end (most recently used)
+            if key in self._cache:
+                self._cache.pop(key)
```

Comment:
With `OrderedDict`, there is a more efficient method `move_to_end` that achieves this effect.
https://docs.python.org/3/library/collections.html#collections.OrderedDict.move_to_end

Reply:
👍